### PR TITLE
Make unwrapCorrupt Check Suppressed Ex.

### LINKED ISSUE #24800: Corruption detection is fragile?
I was looking at how we detect corruptions, and the current logic tries to see whether the exception itself or any of its causes is an instance of `CorruptIndexException`. However, the following pattern that Lucene uses in multiple places:

```java
Throwable priorE = null;
try (ChecksumIndexInput input = dir.openChecksumInput(fileName, context)) {
  // do some stuff that may throw any exceptions
} catch (Throwable exception) {
  priorE = exception;
} finally {
  CodecUtil.checkFooter(input, priorE);
}
```

causes the `CorruptIndexException` to be suppressed if some other exception was thrown in the `try` block. So the thrown exception would not be seen as a corruption exception. This makes me want to check for suppressed exceptions as well in `ExceptionsHelper.unwrapCorruption` but I don't know enough about corruption detection to know whether that is safe to do?

### PR DESCRIPTION
**Pull request by original-brownbear**
* As discussed in #24800 we want to check for suppressed corruption
indicating exceptions here as well to more reliably categorize
corruption related exceptions
* Closes #24800, #41201

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java [hunk 1]
```diff
@@ -438,7 +438,6 @@ protected void failEngine(IOException cause) {
             handler.sendFiles(store, metas.toArray(new StoreFileMetaData[0]), () -> 0);
             fail("exception index");
         } catch (RuntimeException ex) {
-            assertNull(ExceptionsHelper.unwrapCorruption(ex));
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java]: Just removed this as the underlying corruption exception is now unwrapped here propery (it's suppressed on the `RuntimeException`)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

File: server/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java [hunk 2]
```diff
@@ -438,7 +438,7 @@ protected void failEngine(IOException cause) {
...
+            assertNotNull(ExceptionsHelper.unwrapCorruption(ex));
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java]: This now finds the underlying corruption exception that is supressed on the RuntimeException

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

File: server/src/main/java/org/elasticsearch/ExceptionsHelper.java [hunk 3]
```diff
@@ -176,9 +176,17 @@ public static String formatStackTrace(final StackTraceElement[] stackTrace) {
     }
 
     public static IOException unwrapCorruption(Throwable t) {
-        return (IOException) unwrap(t, CorruptIndexException.class,
-                                       IndexFormatTooOldException.class,
-                                       IndexFormatTooNewException.class);
+        IOException corruptionException =
+            (IOException) unwrap(t, CorruptIndexException.class, IndexFormatTooOldException.class, IndexFormatTooNewException.class);
+        if (corruptionException == null && t != null) {
+            for (Throwable suppressed : t.getSuppressed()) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/ExceptionsHelper.java]: do we need to check suppressed exceptions of `Throwable.getCause` etc?

**Review by ywelsch — CHANGES_REQUESTED** [review]
This should also have more tests

File: server/src/main/java/org/elasticsearch/ExceptionsHelper.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/ExceptionsHelper.java]: Right! I guess we should will add that together with tests in a sec :)

File: server/src/main/java/org/elasticsearch/ExceptionsHelper.java [hunk 4]
```diff
@@ -175,10 +175,26 @@ public static String formatStackTrace(final StackTraceElement[] stackTrace) {
         return first;
     }
 
+    private static final List<Class<? extends IOException>> CORRUPTION_EXCEPTIONS =
+        List.of(CorruptIndexException.class, IndexFormatTooOldException.class, IndexFormatTooNewException.class);
+
     public static IOException unwrapCorruption(Throwable t) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/ExceptionsHelper.java]: can you add javadocs to mention that this checks both causes and suppressed exceptions?

Comment (jpountz) [reply server/src/main/java/org/elasticsearch/ExceptionsHelper.java]: and maybe javadocs to `unwrap` too to make it clear it **doesn't** look at suppressed exceptions?

**Review by jpountz — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/ExceptionsHelper.java (hunk 4 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/ExceptionsHelper.java]: Sounds good, added javadoc to both :)

**Review by jpountz — APPROVED** [review]

**Comment by original-brownbear** [comment]
@jpountz thanks
@ywelsch you're ok with this one too? :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
@jpountz @ywelsch thanks!

**Pull request merged**